File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.java
class Solution {
    public List<String> letterCombinations(String digits) {

        List<String> ans = new ArrayList<>();
        if (digits == null || digits.length() == 0) return ans;

        Map<Character, String> mapping = new HashMap<>();
        mapping.put('2', "abc");
        mapping.put('3', "def");
        mapping.put('4', "ghi");
        mapping.put('5', "jkl");
        mapping.put('6', "mno");
        mapping.put('7', "pqrs");
        mapping.put('8', "tuv");
        mapping.put('9', "wxyz");

        backtrack(0,ans,"", digits, mapping);
        return ans;
    }

    private void backtrack(int index,List<String> ans, String current, String digits, Map<Character, String> mapping) {
        if (index == digits.length()) {
            ans.add(current);
            return;
        }

        String letters = mapping.get(digits.charAt(index));
        for (char ch : letters.toCharArray()) {
            backtrack(index + 1,ans,current + ch, digits, mapping);
        }
    }
}

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.py
# from typing import List

# class Solution:
#     def letterCombinations(self, digits: str) -> List[str]:
#         if not digits:
#             return []

#         mapping = {
#             "2": "abc",
#             "3": "def",
#             "4": "ghi",
#             "5": "jkl",
#             "6": "mno",
#             "7": "pqrs",
#             "8": "tuv",
#             "9": "wxyz"
#         }

#         ans = []

#         def backtrack(index: int, current: str):
#             if index == len(digits):
#                 ans.append(current)
#                 return

#             for ch in mapping[digits[index]]:
#                 backtrack(index + 1, current + ch)

#         backtrack(0, "")
#         return ans
from typing import List
from itertools import product

class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if not digits:
            return []

        mapping = {
            "2": "abc",
            "3": "def",
            "4": "ghi",
            "5": "jkl",
            "6": "mno",
            "7": "pqrs",
            "8": "tuv",
            "9": "wxyz"
        }

        # Build the list of letters for each digit
        groups = [mapping[d] for d in digits]

        # Cartesian product of all groups
        return ["".join(p) for p in product(*groups)]
